src/lib.rs

```diff
@@ -51,144 +49,406 @@
-);
+use std::{
+    collections::HashSet,
+    env, fs,
+    io::{Read, Write},
+    path::Path,
+    path::PathBuf,
+};
+
+use chrono::{DateTime, Duration, Utc};
+use getrandom::getrandom;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+pub mod prelude {
+    pub use crate::clients::{BaseClient, OAuthClient};
+}
+
+pub(in crate) mod headers {
+    // Common headers as constants
+    pub const CLIENT_ID: &str = "client_id";
+    pub const CODE: &str = "code";
+    pub const GRANT_AUTH_CODE: &str = "authorization_code";
+    pub const GRANT_CLIENT_CREDS: &str = "client_credentials";
+    pub const GRANT_REFRESH_TOKEN: &str = "refresh_token";
+    pub const GRANT_TYPE: &str = "grant_type";
+    pub const REDIRECT_URI: &str = "redirect_uri";
+    pub const REFRESH_TOKEN: &str = "refresh_token";
+    pub const RESPONSE_CODE: &str = "code";
+    pub const RESPONSE_TYPE: &str = "response_type";
+    pub const SCOPE: &str = "scope";
+    pub const SHOW_DIALOG: &str = "show_dialog";
+    pub const STATE: &str = "state";
+    // TODO:
+    // pub const CODE_CHALLENGE: &str = "code_challenge";
+    // pub const CODE_CHALLENGE_METHOD: &str = "code_challenge_method";
+}
+
+pub(in crate) mod auth_urls {
+    pub const AUTHORIZE: &str = "https://accounts.spotify.com/authorize";
+    pub const TOKEN: &str = "https://accounts.spotify.com/api/token";
+}
+
+/// Possible errors returned from the `rspotify` client.
+#[derive(Debug, Error)]
+pub enum ClientError {
+    #[error("json parse error: {0}")]
+    ParseJson(#[from] serde_json::Error),
+
+    #[error("url parse error: {0}")]
+    ParseUrl(#[from] url::ParseError),
+
+    #[error("http error: {0}")]
+    Http(#[from] http::Error),
+
+    #[error("input/output error: {0}")]
+    Io(#[from] std::io::Error),
+
+    #[cfg(feature = "cli")]
+    #[error("cli error: {0}")]
+    Cli(String),
+
+    #[error("cache file error: {0}")]
+    CacheFile(String),
+}
+
+pub type ClientResult<T> = Result<T, ClientError>;
+
+pub const DEFAULT_API_PREFIX: &str = "https://api.spotify.com/v1/";
+pub const DEFAULT_CACHE_PATH: &str = ".spotify_token_cache.json";
+pub const DEFAULT_PAGINATION_CHUNKS: u32 = 50;
+
+/// Struct to configure the Spotify client.
+#[derive(Debug, Clone)]
+pub struct Config {
+    /// The Spotify API prefix, [`DEFAULT_API_PREFIX`] by default.
+    pub prefix: String,
+
+    /// The cache file path, in case it's used. By default it's
+    /// [`DEFAULT_CACHE_PATH`]
+    pub cache_path: PathBuf,
+
+    /// The pagination chunk size used when performing automatically paginated
+    /// requests, like [`artist_albums`](crate::clients::BaseClient). This
+    /// means that a request will be performed every `pagination_chunks` items.
+    /// By default this is [`DEFAULT_PAGINATION_CHUNKS`].
+    ///
+    /// Note that most endpoints set a maximum to the number of items per
+    /// request, which most times is 50.
+    pub pagination_chunks: u32,
+
+    /// TODO
+    pub token_cached: bool,
```

Comment:
I think `token_cached` is implemented to control whether to save the token cache, so it's fine to remove the `TODO` mark

Reply:
Done, thanks.